contracts/non-fungible-tokens/andromeda-ics721/src/contract.rs

```diff
@@ -0,0 +1,619 @@
+    env: Env,
+    info: MessageInfo,
+    receiver: String,
+    class_uri: Option<String>,
+    transfers: Option<TransferInfo>,
+    new_tokens: Option<NewTokenInfo>,
+) -> Result<Response, ContractError> {
+    ensure!(
+        info.sender == env.contract.address,
+        ContractError::Unauthorized {}
+    );
+
+    let receiver = deps.api.addr_validate(&receiver)?;
+
+    let mut messages: Vec<WasmMsg> = Vec::with_capacity(2);
+    if let Some(transfer_info) = transfers {
+        messages.push(transfer_info.into_wasm_msg(&env, &receiver)?)
+    }
+    if let Some(token_info) = new_tokens {
+        messages.push(token_info.into_wasm_msg(&env, &receiver, class_uri)?)
+    }
+
+    Ok(Response::default()
+        .add_attribute("method", "handle_packet_receive")
+        .add_messages(messages))
+}
+
+fn execute_mint(
+    deps: Deps,
+    env: Env,
+    info: MessageInfo,
+    class_id: String,
+    token_ids: Vec<String>,
+    token_uris: Vec<String>,
+    receiver: String,
+) -> Result<Response, ContractError> {
+    ensure!(
+        info.sender == env.contract.address,
+        ContractError::Unauthorized {}
+    );
+    ensure!(
+        token_ids.len() == token_uris.len(),
+        ContractError::ImbalancedTokenInfo {}
+    );
+
+    let receiver = deps.api.addr_validate(&receiver)?;
+    let cw721_addr = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class_id)?;
+
+    let mint_messages = token_ids
+        .into_iter()
+        // Can zip without worrying about dropping data as we assert
+        // that lengths are the same above.
+        .zip(token_uris.into_iter())
+        .map(|(token_id, token_uri)| -> StdResult<WasmMsg> {
+            let msg = cw721_base::msg::ExecuteMsg::<Empty, Empty>::Mint(cw721_base::MintMsg {
+                token_id,
+                token_uri: Some(token_uri),
+                owner: receiver.to_string(),
+                extension: Empty::default(),
+            });
+            Ok(WasmMsg::Execute {
+                contract_addr: cw721_addr.to_string(),
+                msg: to_binary(&msg)?,
+                funds: vec![],
+            })
+        })
+        .collect::<StdResult<Vec<_>>>()?;
+
+    Ok(Response::default()
+        .add_attribute("method", "execute_mint")
+        .add_messages(mint_messages))
+}
+
+#[allow(clippy::too_many_arguments)]
+fn execute_do_instantiate_and_mint(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    class_id: String,
+    class_uri: Option<String>,
+    token_ids: Vec<String>,
+    token_uris: Vec<String>,
+    receiver: String,
+) -> Result<Response, ContractError> {
+    ensure!(
+        info.sender == env.contract.address,
+        ContractError::Unauthorized {}
+    );
+
+    // Optionally, instantiate a new cw721 contract if one does not
+    // yet exist.
+    let submessages = if CLASS_ID_TO_NFT_CONTRACT.has(deps.storage, class_id.clone()) {
+        vec![]
+    } else {
+        let message = SubMsg::<Empty>::reply_on_success(
```

Comment:
What happens if the instantiation fails?

Reply:
The whole operation is supposed to rollback upon an error

Reply:
Reply On Success ignores if the sub messages fails, is it erroring before the end of this method?